Implement message translation for multiple languages, change display language and translate messages.

hoangdinhphi92 opened the pull request:
Based on [issue](https://github.com/suesitran/public_chat/issues/12), I have made the following changes:

1. Modified firebase functions to add languages supported by Gemini according to [the table](https://support.google.com/gemini/answer/14579026?visit_id=638666714290548199-1329690995&p=gemini_app_requirements&rd=1)

2. Added settings screen including two screens:
- Screen to change in-app language based on the list of supported languages from AppLocalizations.supportedLocales
- Screen to change translation language based on supported languages list returned from firebase functions. Default is device language or English.

3. Fixed several UI-related bugs.

[Example Video.](https://youtu.be/LMqY7cld38c)

I have followed the Bloc pattern you implemented. Please let me know if there are any issues that need to be addressed. Thank you.

thenartus commented:
Good implementation, with tests

suesitran commented on lib/_shared/widgets/chat_bubble_widget.dart:
```diff
@@ -10,96 +12,110 @@ class ChatBubble extends StatelessWidget {
+                    )
+            ],
           ),
-          // english version (if there is)
```
if this English translation is deleted, where on the UI does the translation show?

hoangdinhphi92 replied:
This is original comment code not relate my edited code, so I will change this comment.

suesitran commented on lib/features/genai_setting/bloc/genai_bloc.dart:
```diff
@@ -7,6 +7,7 @@ import 'package:public_chat/service_locator/service_locator.dart';
 import 'package:public_chat/utils/bloc_extensions.dart';
```
changes in this file is not related to this issue

hoangdinhphi92 replied:
When I tested the app I didn't see any loading state while waiting for Gemini response so I added this feature. I will remove it now and create a new PR when needed

suesitran commented on lib/features/settings/chat_language/service/chat_language_service.dart:
```diff
@@ -0,0 +1,19 @@
+    try {
+      final result =
+          await _functions.httpsCallable('getSupportedLanguages').call();
+      final List<dynamic> json = jsonDecode(jsonEncode(result.data));
```
why do u need to encode a json, then decode it at the same time?

hoangdinhphi92 replied:
I already use cloud function before. Sometime I has issue relate json output type like this [issue](https://github.com/firebase/flutterfire/issues/4059) . So I need decode and encode same time to correct json type. It's just a habit when working with cloud functions to make sure there are no bugs.

suesitran commented on lib/features/settings/settings_screen.dart:
```diff
@@ -0,0 +1,110 @@
+import 'package:flutter/material.dart';
```
break App locale setting to another PR.

Each PR should fix only 1 issue.

hoangdinhphi92 replied:
To be honest, at first I thought this issue was related to changing the language of the application, so I did it first. Then when I looked at it specifically I realized I was wrong. Instead of deleting it, I added it to this PR. I will separate it to make sure the PR only fixes 1 issue. Thank you,

suesitran reviewed (changes requested):
good PR overall, please fix my comments then i'll review again

hoangdinhphi92 commented on functions/index.js:
```diff
@@ -34,62 +35,81 @@ const generationConfig = {
+
     const curTranslated = document["translated"];
 
-    // check if message is translated
```
This is my mistake. I will add comment again.

hoangdinhphi92 commented on lib/features/settings/locale/data/locale_info.dart:
```diff
@@ -0,0 +1,75 @@
+const Map<String, Map<String, String>> kLocaleData = {
```
Language config in cloud functions for chat translation. It maybe not same with application language. When these two lists differ, correction is difficult and time consuming.

hoangdinhphi92 commented:
I made a few changes:
- Added comments to the cloud functions code.
- Removed the app language change feature, will create a PR after this PR can be merged.
- Removed the loading feature while waiting for Gemini response.

I hope you will review and give me feedback if there is any problem. Thank you.

thenartus commented:
Please send me messenger message 👍

hoangdinhphi92 commented:
> Gửi tin nhắn messenger cho chị nha

Facebook của em là Hoàng Phi nha chị.
